From 634031a7e846d12c8cc6e979521dcb291f550a83 Mon Sep 17 00:00:00 2001
Message-Id: <634031a7e846d12c8cc6e979521dcb291f550a83.1429347747.git.chang-joon.lee@intel.com>
In-Reply-To: <58b01df1cceca63ad29c5d3026fe071bcc8e2015.1429347747.git.chang-joon.lee@intel.com>
References: <58b01df1cceca63ad29c5d3026fe071bcc8e2015.1429347747.git.chang-joon.lee@intel.com>
From: N Tajunnisha <tajunnisha.n@intel.com>
Date: Wed, 15 Apr 2015 00:00:23 +0530
Subject: [PATCH 3/4] REVERTME [VPG]: drm/i915: Adding X-Power PMIC support

GOP currently supports TI PMIC for CHT CR T3 boards.
Android has Xpower PMIC as POR. MIPI Blankout is observed
because of using wrong PMIC values from vbt for PMIC
programming during DSI enable and disable.
This patch is hardcoding appropriate PMIC values for
Xpower PMIC configuration till this support gets
added in BIOS and GOP.

Change-Id: I802732b7a704a6ec61f3aeb995cb8f41d509e424
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-25164
Signed-off-by: N Tajunnisha <tajunnisha.n@intel.com>
Signed-off-by: Uma Shankar <uma.shankar@intel.com>
---
 drivers/gpu/drm/i915/i915_reg.h            |    8 ++++++++
 drivers/gpu/drm/i915/intel_dsi_panel_vbt.c |   29 +++++++++++++++++++++-------
 2 files changed, 30 insertions(+), 7 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index b273796..4b08057 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -1797,6 +1797,11 @@ enum punit_power_well {
 #define   FBC_REND_NUKE		(1<<2)
 #define   FBC_REND_CACHE_CLEAN	(1<<1)
 
+/* PMIC Registers */
+#define XPOWER_PMIC_PANEL_POWER_CTRL_REG 0x12
+#define XPOWER_PMIC_PANEL_POWER_CTRL_DATAMASK (1 << 6)
+#define XPOWER_PMIC_PANEL_POWER_ON_DATA (1 << 6)
+
 /*
  * GPIO regs
  */
@@ -7086,4 +7091,7 @@ enum punit_power_well {
 
 #define GEN8_RC6_WA_BB	    0x2058
 
+/* CHT_CR Related */
+#define CHT_CR_REVISION 0x22
+
 #endif /* _I915_REG_H_ */
diff --git a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
index d8d4065..09a5202 100644
--- a/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
+++ b/drivers/gpu/drm/i915/intel_dsi_panel_vbt.c
@@ -545,6 +545,7 @@ static u8 *mipi_exec_spi(struct intel_dsi *intel_dsi, u8 *data)
 
 static u8 *mipi_exec_pmic(struct intel_dsi *intel_dsi, u8 *data)
 {
+	struct drm_device *dev = intel_dsi->base.base.dev;
 	u32 register_address, register_data;
 	int data_mask, tmp;
 	int ret;
@@ -553,14 +554,28 @@ static u8 *mipi_exec_pmic(struct intel_dsi *intel_dsi, u8 *data)
 	 * First 3 bytes are not relevant for Linux.
 	 * Skipping the data field by 3 bytes to get
 	 * the PMIC register Address.
+	 * For CHT-CR, Xpower PMIC is POR for Android
+	 * but GOP supports a different PMIC for windows.
+	 * So, hardcoding the PMIC register and values for now
+	 * until GOP supports XPower PMIC.
 	 */
-	data += 3;
-	register_address = *((u32 *)data);
-	data += 4;
-	register_data = *((u32 *)data);
-	data += 4;
-	data_mask = *((u32 *)data);
-	data += 4;
+	if (IS_CHERRYVIEW(dev) && dev->pdev->revision == CHT_CR_REVISION) {
+		data += 7;
+		register_address = XPOWER_PMIC_PANEL_POWER_CTRL_REG;
+		register_data = *((u32 *)data);
+		data_mask = XPOWER_PMIC_PANEL_POWER_CTRL_DATAMASK;
+		if (register_data != 0)
+			register_data = XPOWER_PMIC_PANEL_POWER_ON_DATA;
+		data += 8;
+	} else {
+		data += 3;
+		register_address = *((u32 *)data);
+		data += 4;
+		register_data = *((u32 *)data);
+		data += 4;
+		data_mask = *((u32 *)data);
+		data += 4;
+	}
 
 	tmp = intel_soc_pmic_readb(register_address);
 	if (tmp < 0) {
-- 
1.7.9.5

